### PR TITLE
Tzr

### PR DESCRIPTION
**Pull request by luojing1211**
This pull request implements the TZR absolute phase. The unit test is not provided yet.

### CODE REVIEW

**Comment by luojing1211** [comment]
@scottransom @paulray Could you review this PR. @paulray do you have a simple test case for this model?

**Comment by paulray** [comment]
I'll compare it to my little hack in photonphase and let you know.

File: pint/models/timing_model.py [hunk 1]
```diff
@@ -477,15 +477,31 @@ def delay(self, toas, cutoff_component='', include_last=True):
...
         phase = Phase(np.zeros(len(toas)) , np.zeros(len(toas)))
         # Then compute the relevant pulse phases
         for pf in self.phase_funcs:
             phase += Phase(pf(toas, delay))
-        return phase
+
+        # If the absolute phase flag is on, use the TZR parameters to compute
+        # the absolute phase.
+        if abs_phase:
+            if 'AbsPhase' not in list(self.component.keys()):
```
Comment (paulray) [inline pint/models/timing_model.py]: This  should be self.components.keys()
(It is missing the s in components)

Comment (paulray) [inline pint/models/timing_model.py]: Also, I loaded a par file using pint.models.get_model().  Even though it has TZR parameters, it is not adding an AbsPhase component.  Here is what is has:
```
In [3]: m.components
Out[3]: 
{'AstrometryEquatorial': <pint.models.astrometry.AstrometryEquatorial at 0x10ac87c50>,
 'Dispersion': <pint.models.dispersion_model.Dispersion at 0x10ac7bf10>,
 'SolarSystemShapiro': <pint.models.solar_system_shapiro.SolarSystemShapiro at 0x10ac96650>,
 'Spindown': <pint.models.spindown.Spindown at 0x10ac74450>}
```

**Comment by paulray** [comment]
Still some bugs.
```
  File "/Users/paulr/src/PINT/pint/models/__init__.py", line 16, in <module>
    from .absolute_phase import AbsPhase
  File "/Users/paulr/src/PINT/pint/models/absolute_phase.py", line 57
    tz = toa.get_TOAs_list([TZR_toa,], include_bipm=clkc_info{'include_bipm'},
```
I think the curly braces should be brackets.

**Comment by paulray** [comment]
Next, I think that `abs_phase=True` should be the default if there are TZR parameters in the model (i.e. if the AbsPhase component exists).

**Comment by luojing1211** [comment]
Thank you, I am fixing it. Is the overall structure ok?

**Comment by luojing1211** [comment]
I just realized passing the TOA information (ephemeris, clock correction flag... ) into the AbsPhase class requires changing the API of phase(). Currently, phase() accepts the TOAs.table. However, the setup of this PR requires the TOAs class. Is there any other way of avoiding it?

**Comment by paulray** [comment]
Yeah, I've never liked having to figure out which functions want TOAs and which want TOAs.table. Having to pass TOAs.table really breaks the encapsulation of the TOAs class.  But, short of reversing that everywhere, I'm not sure the best solution in this case.

**Comment by luojing1211** [comment]
The alternative is to append the TZR toa to the TOAs table and recompute the TDB and positions for all the TOAs again.

**Comment by paulray** [comment]
Computing TDBs and positions is expensive, so I wouldn't want to do that again for all the TOAs.
I think @demorest was an advocate for passing TOAs.table (and maybe @scottransom too, I'm not sure), but it does lose metadata, as you point out. Perhaps we could revisit that discussion?

**Comment by scottransom** [comment]
What would be the problem with changing everything to accept the TOAs class rather than just passing the table?  I don't think it would slow things down, given that we could do something like tbl = toa.table on the first line of each function that uses the table currently.

**Comment by demorest** [comment]
I don't remember what (if any) rationale there was for passing the table rather than the TOAs object.  It may just be due to the history of the code.. I think we were originally simply using array-valued TDB Times for these arguments.  That has a nice simplicity but at some point I guess we needed more info from the table.  I agree that using the TOAs class makes more sense and could allow us to work towards something where the table is "hidden" inside TOAs and never directly accessed by outside code.  This makes it a bit more modular in case we want to change how the info is stored inside TOAs.

**Comment by paulray** [comment]
Other than an API change, are there any negatives to changing all routines to accept `TOAs` instead of `TOAs.table`?  Since it is an externally visible API change, it is definitely good to do before version 1.0 (and before the paper is finalized), if we are going to do it, so we don't have that major inconsistency floating around.

**Comment by scottransom** [comment]
I can't think of any issues that would be caused by passing a TOAs class instance rather than the toa table.  And I do think it is cleaner to do that.  So I think we should probably just make the changes (and put that in the paper if necessary).

**Comment by luojing1211** [comment]
Ok, if there is no conflict we can think of, I will just start to change it.

**Pull request merged**